### ISSUE TITLE
Numerical instabilities with LoHa

### ISSUE THREAD

**Issue by BenjaminBossan** [huggingface/peft#1058]
### System Info

Latest PEFT main.

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import torch
from torch import nn
from peft import LoraConfig, LoHaConfig, get_peft_model

class MLP(nn.Module):
    def __init__(self, bias=True):
        super().__init__()
        self.lin0 = nn.Linear(10, 20, bias=bias)
        self.relu = nn.ReLU()
        self.drop = nn.Dropout(0.5)
        self.lin1 = nn.Linear(20, 2, bias=bias)
        self.sm = nn.LogSoftmax(dim=-1)

    def forward(self, X):
        X = X.float()
        X = self.lin0(X)
        X = self.relu(X)
        X = self.drop(X)
        X = self.lin1(X)
        X = self.sm(X)
        return X

def test_it(config_cls, device):
    input = torch.arange(90).reshape(9, 10).to(device)
    torch.manual_seed(0)
    if issubclass(config_cls, LoraConfig):
        config = config_cls(target_modules=["lin1"], init_lora_weights=False)
    else:
        config = config_cls(target_modules=["lin1"], init_weights=False)
    peft_model = get_peft_model(MLP(), config).eval().to(device)
    output = peft_model(input)
    is_finite = torch.isfinite(output).all().item()
    return is_finite

if __name__ == "__main__":
    n = 100
    print("LoHa")

    device = "cpu"
    print("Testing on CPU")
    successes = [test_it(LoHaConfig, device) for _ in range(n)]
    print(f"Success rate: {sum(successes) / n}")

    device = "cuda"
    print("\nTesting on GPU")
    successes = [test_it(LoHaConfig, device) for _ in range(n)]
    print(f"Success rate: {sum(successes) / n}")

    print("Lora")

    device = "cpu"
    print("Testing on CPU")
    successes = [test_it(LoraConfig, device) for _ in range(n)]
    print(f"Success rate: {sum(successes) / n}")

    device = "cuda"
    print("\nTesting on GPU")
    successes = [test_it(LoraConfig, device) for _ in range(n)]
    print(f"Success rate: {sum(successes) / n}")
```

### Expected behavior

While running some tests that involved LoHa, I came across the issue that they were failing randomly. For some strange reason, LoHa seems to be numerically unstable, as running the attached script should show. It applies LoHa on a simple MLP and checks if the output is finite or not (if not, it's NaNs). For control, I run on GPU vs CPU and with LoHa vs LoRA:

```
LoHa
Testing on CPU
Success rate: 0.21
Testing on GPU
Success rate: 0.25

Lora
Testing on CPU
Success rate: 1.0
Testing on GPU
Success rate: 1.0
```

Even more strangely, the rate at which these instabilities occurs varies considerably, much more than should be expected by chance, e.g. sometimes I get (omitting LoRA, which always succeeds):

```
LoHa
Testing on CPU
Success rate: 0.21
Testing on GPU
Success rate: 0.25
```

And other times, I get:

```
LoHa
Testing on CPU
Success rate: 0.98
Testing on GPU
Success rate: 0.6
```

So the success rate for CPU can vary from 21/100 to 98/100! Chi² test gives a p< 0.00001 (uncorrected), so this cannot be an accident.

Can anyone reproduce?

Edit: Some more tests: 

1. Using `torch.randn(9, 10)` as input doesn't help much.
2. Targeting `["lin0"]` or `["lin0", "lin1"]` makes matters worse.
3. `r` doesn't seem to make a difference.

**Comment by BenjaminBossan** [comment]
ping @kovalexal

**Comment by kovalexal** [comment]
@BenjaminBossan, thanks for bringing it up.

I am wondering, why are you running your test with `init_weights=False`?
IMHO `init_weights=False` should be used only in the situation where you are trying to construct an empty adapter to load from a checkpoint later on (so, you don't want to spend time to initialize weights properly).

Under the hood, when you pass `init_weights=False` current [LoHa](https://github.com/huggingface/peft/blob/207229ad5e1854cd897306ce5ee725b0245b1064/src/peft/tuners/loha/layer.py#L70) and [LoKr](https://github.com/huggingface/peft/blob/f6e73352197d309382b964c28fa764d8caec2a2f/src/peft/tuners/lokr/layer.py#L73) implementations create adapter tensors with [`torch.empty`](https://pytorch.org/docs/stable/generated/torch.empty.html), which provides uninitialized memory buffers, so `torch.randn` does not make a difference in that case.

Here is a small demo that showcases what is happening under the hood without all the adapters logic:

```python
import torch

def test_finite(device):
    w1a = torch.empty((2, 8)).to(device=device)
    w1b = torch.empty((8, 20)).to(device=device)
    w2a = torch.empty((2, 8)).to(device=device)
    w2b = torch.empty((8, 20)).to(device=device)

    # LoHa way for contructing delta
    w1a_w1b = w1a @ w1b
    w2a_w2b = w2a @ w2b
    delta = w1a_w1b * w2a_w2b

    return {
        "w1a": torch.isfinite(w1a).all().item(),
        "w1b": torch.isfinite(w1b).all().item(),
        "w2a": torch.isfinite(w1a).all().item(),
        "w2b": torch.isfinite(w1b).all().item(),
        "w1a_w1b": torch.isfinite(w1a_w1b).all().item(),
        "w2a_w2b": torch.isfinite(w2a_w2b).all().item(),
        "delta": torch.isfinite(delta).all().item(),
    }


n = 10000

# CPU results
cpu_results = [test_finite("cpu") for _ in range(n)]
cpu_results_combined = cpu_results[0]
for result in cpu_results[1:]:
    for key, value in result.items():
        cpu_results_combined[key] += value
print("Success rate CPU:")
for key, value in cpu_results_combined.items():
    print(f"{key}: {value / n}")


# GPU results
gpu_results = [test_finite("cuda:0") for _ in range(n)]
gpu_results_combined = gpu_results[0]
for result in gpu_results[1:]:
    for key, value in result.items():
        gpu_results_combined[key] += value
print("Success rate GPU:")
for key, value in gpu_results_combined.items():
    print(f"{key}: {value / n}")
```

Here is a sample output:
```
Success rate CPU:
Success rate CPU:
w1a: 0.9969
w1b: 0.681
w2a: 0.9969
w2b: 0.681
w1a_w1b: 0.6638
w2a_w2b: 0.6148
delta: 0.4431

Success rate GPU:
w1a: 0.9997
w1b: 0.6242
w2a: 0.9997
w2b: 0.6242
w1a_w1b: 0.6236
w2a_w2b: 0.6701
delta: 0.4167
```

Also, your sample starts to work as expected when you pass `init_weights=True` to LoHa.

Maybe I am missing some logic in terms of `init_weights` parameter, could you please clarify, how it should be processed by default when `False` is passed?

**Comment by BenjaminBossan** [comment]
> Under the hood, when you pass `init_weights=False` current [LoHa](https://github.com/huggingface/peft/blob/207229ad5e1854cd897306ce5ee725b0245b1064/src/peft/tuners/loha/layer.py#L70) and [LoKr](https://github.com/huggingface/peft/blob/f6e73352197d309382b964c28fa764d8caec2a2f/src/peft/tuners/lokr/layer.py#L73) implementations create adapter tensors with [`torch.empty`](https://pytorch.org/docs/stable/generated/torch.empty.html)

Thanks, that explains the strange pattern I saw.

> Maybe I am missing some logic in terms of `init_weights` parameter, could you please clarify, how it should be processed by default when `False` is passed?

It is indeed confusing, I also didn't understand it at first. The idea for `init_weights=True` is of course the initialize the weights as described by the paper. Often, this means that they are initialized in a way to be the identity transform, as that facilitates training.

When we do testing, we want the adapter _not_ to be the identity transform, because that would mean that the results are the same as without adapter, i.e. there is no effect. Therefore, the possibility to set `init_weights=False` was added. It is basically a setting for debugging/testing. Of course, we could set `init_weights=True` and then train the adapter so  that it has an effect, but this is a lot of extra work and makes testing slower and more verbose. This is why a lot of tests use `init_weights=False`.

Note that for LoRA, we do _not_ initialize the LoRA weights with empty tensors when using `init_lora_weights=False`:

https://github.com/huggingface/peft/blob/207229ad5e1854cd897306ce5ee725b0245b1064/src/peft/tuners/lora/layer.py#L78-L79

> IMHO `init_weights=False` should be used only in the situation where you are trying to construct an empty adapter to load from a checkpoint later on

This is a good point. Probably it's not a huge time saver, as the adapter weights should typically be small, but could still be nice to have. I wonder if we can add it as a third option.
